### PR TITLE
refactor file and add s3 support for CaffePersister

### PR DESCRIPTION
**Pull request by qiuxin2012**
## What changes were proposed in this pull request?

refactor file and add s3 support for CaffePersister

## How was this patch tested?

unit tests, integration tests, manual tests

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/File.scala [hunk 1]
```diff
@@ -17,16 +17,70 @@
...
+/**
+ * A uniform approach to get Input/Output stream in BigDL.
+ * @param fileName
+ */
+private[bigdl] class File(fileName: String) {
+  private val inputStreams: ArrayBuffer[InputStream] = new ArrayBuffer[InputStream]()
+  private var outputStream: OutputStream = null
+  private val conf = File.getConfiguration(fileName)
+  private val path = new Path(fileName)
+  private val fs: FileSystem = path.getFileSystem(conf)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/File.scala]: seems this piece of code specifically for HDFS ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/File.scala [hunk 2]
```diff
@@ -17,16 +17,70 @@
...
 
+import scala.collection.mutable.ArrayBuffer
+
+/**
+ * A uniform approach to get Input/Output stream in BigDL.
+ * @param fileName
+ */
+private[bigdl] class File(fileName: String) {
+  private val inputStreams: ArrayBuffer[InputStream] = new ArrayBuffer[InputStream]()
+  private var outputStream: OutputStream = null
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/File.scala]: why do we need a declared variable ? I didn't see any usage inside the class  (just assigin a created output stream to it and close it), the create method below just return the outoutstream to external reference and we could close it there as you do it now

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/File.scala [hunk 3]
```diff
@@ -17,16 +17,70 @@
...
 import org.apache.hadoop.io.IOUtils
 
+import scala.collection.mutable.ArrayBuffer
+
+/**
+ * A uniform approach to get Input/Output stream in BigDL.
+ * @param fileName
+ */
+private[bigdl] class File(fileName: String) {
+  private val inputStreams: ArrayBuffer[InputStream] = new ArrayBuffer[InputStream]()
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/File.scala]: do we have the scenario we need to maintain multiple input stream for the same file ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/File.scala (hunk 1 above)
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/File.scala]: also works for local file system and s3.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/caffe/CaffePersister.scala [hunk 4]
```diff
@@ -211,9 +211,9 @@ class CaffePersister[T: ClassTag](val prototxtPath: String,
       })
     }
 
-    var prototxtFile: File = null
+    var prototxtFile: FileWriter = null
     try {
-      prototxtFile = File(prototxtPath)
+      prototxtFile = FileWriter(prototxtPath)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/caffe/CaffePersister.scala]: Generally change the prototxtFile to prototxtFileWriter would make more sense ?

**Comment by wzhongyuan** [comment]
@qiuxin2012   this generally looks good to me, only thing that should we modify a little bit on the naming like from ***file to ***fileWriter(Reader) as we separate reading and writing to different objects
thanks

**Comment by wzhongyuan** [comment]
LGTM

**Comment by qiuxin2012** [comment]
jenkins passed

**Pull request merged**